tud-zih-energy/otf2xx
# Use TMP to simplify code

Based on #32. This shows the strengths of the approach added there: Using a unique `reference<>` type per ID namespace
  - With this it was discovered, that `io_pre_created_handle_state` had its own references which is wrong. This was fixed.

New:
  - Add basic TMP (template meta programming): concat, contains, apply, transform
  - Test those
  - Add lists `referable_definitions_base`, `referable_definitions_ext`, `unreferable_definitions` and sort definitions into those (not including templated ones)
  - Use this to automatically construct a tuple of reference generators for all base definitions (id spaces) which removes 90% of the code in `reference_generator.hpp`

The same can be used for the registry to get rid of all the duplicate code there.

Note how this also provides meaningful meta data on definitions that can be used further. E.g. it should be impossible to create a `reference<>` of a type not in `referable_definitions_base`.
Note also, how it makes adding a new definition as easy as adding it to a list. Previously it required changing code in multiple locations which can easily be forgotten. (3x in reference_generator.hpp, ~5 in registry.hpp)
